### PR TITLE
Materials class and temperature processing

### PR DESCRIPTION
**Pull request by jhdark**
## Proposed changes

Introducing the Material class, this will be an argument of a VolumeSubdomain class and will define certain hydrogen transport properties within a given subdomain. Additionally, some minor processing of the temperature argument is added to ensure its in the correct format for the formulation. More complex temperature arguments to come in the future.

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [x] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: test/test_material.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    assert float(D) == D_analytical
+
+
+@pytest.mark.parametrize(
+    "test_type", [1, fem.Constant(test_mesh.mesh, 1.0), 1.0, "coucou", 2 * x[0]]
+)
+def test_diffusion_values_type(test_type):
+    """Test that the diffusion coefficient values types are correctly
+    processed"""
+    my_mat = F.Material(test_type, test_type, "1")
```
Comment (RemDelaporteMathurin) [inline test/test_material.py:29]:
```suggestion
    my_mat = F.Material(D_0=test_type, E_D=test_type, name="1")
```

File: test/test_material.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    if isinstance(test_type, (fem.Constant, int, float)):
+        my_mat.define_diffusion_coefficient(
+            test_mesh.mesh, fem.Constant(test_mesh.mesh, 1.0)
+        )
+    else:
+        with pytest.raises(TypeError):
+            my_mat.define_diffusion_coefficient(
+                test_mesh.mesh, fem.Constant(test_mesh.mesh, 1.0)
+            )
```
Comment (RemDelaporteMathurin) [inline test/test_material.py:38]: I would replace `test_type` by `value`

Comment (RemDelaporteMathurin) [inline test/test_material.py:38]: I think that once you have a setter for D_0 and E_D you will be able to rewrite this test into two tests:

```python
@pytest.mark.parametrize(
    "value", [1, fem.Constant(test_mesh.mesh, 1.0), 1.0, "coucou", 2 * x[0]]
)
def test_mat_D_0_type(value):
    if isinstance(test_type, (fem.Constant, int, float)):
        my_mat = F.Material(D_0=value, E_D=0, name="mat")
        assert isinstance(my_mat.D_0, fem.Constant)
    else:
        with pytest.raises(TypeError):
            my_mat = F.Material(D_0=value, E_D=0, name="mat")

@pytest.mark.parametrize(
    "value", [1, fem.Constant(test_mesh.mesh, 1.0), 1.0, "coucou", 2 * x[0]]
)
def test_mat_E_D_type(value):
    if isinstance(test_type, (fem.Constant, int, float)):
        my_mat = F.Material(D_0=1, E_D=value, name="mat")
        assert isinstance(my_mat.E_D, fem.Constant)
    else:
        with pytest.raises(TypeError):
            my_mat = F.Material(D_0=1, E_D=value, name="mat")
```

File: test/test_temperature.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    "test_value",
+    [1, fem.Constant(test_mesh.mesh, 1.0), 1.0],
+)
+def test_temperature_value_processing(test_value):
+    """Test that the temperature type is correctly processed"""
+    my_model = F.HydrogenTransportProblem()
+    my_model.mesh = test_mesh
+    my_model.temperature = test_value
+
+    assert isinstance(my_model.temperature, fem.Constant)
```
Comment (RemDelaporteMathurin) [inline test/test_temperature.py:30]: Why not merge these two tests?

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
First round of comments

File: festim/hydrogen_transport_problem.py [hunk 4]
```diff
@@ -88,6 +88,22 @@ def __init__(
...
+    @temperature.setter
+    def temperature(self, value):
+        if value is None:
+            self._temperature = value
+        elif isinstance(value, (float, int, fem.Constant)):
+            self._temperature = F.as_fenics_constant(value, self.mesh.mesh)
+        else:
+            raise TypeError(
+                f"Temperature must be float or dolfinx.Constant, not {type(value)}"
+            )
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py:106]: You could refactor this by using the new `as_fenics_constant` function

File: festim/hydrogen_transport_problem.py [hunk 5]
```diff
@@ -159,39 +176,38 @@ def create_formulation(self):
...
-            # for bc in self.boundary_conditions:
-            #     pass
-            #     if bc.species == spe and bc.type != "dirichlet":
-            #         formulation += bc * v * self.ds
-
-        self.formulation = formulation
+            for vol in self.volume_subdomains:
+                u = spe.solution
+                u_n = spe.prev_solution
+                v = spe.test_function
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py:190]: No need to redefine the functions for all subdomains only for each species
```suggestion
            u = spe.solution
            u_n = spe.prev_solution
            v = spe.test_function
            for vol in self.volume_subdomains:
```

File: festim/material.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import ufl
+import festim as F
+
+
+class Material:
+    """
+    Material class
+    """
```
Comment (RemDelaporteMathurin) [inline festim/material.py:8]: Please put arguments and attributes here

File: festim/material.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+    def __init__(self, D_0, E_D, name=None) -> None:
+        """Inits Material
+        Args:
+            D_0 (float or fem.Constant): the diffusion coefficient at 0 K
+            E_D (float or fem.Constant): the activation energy for diffusion
+            name (str): the name of the material
+        """
```
Comment (RemDelaporteMathurin) [inline festim/material.py:16]: See comment above

File: test/test_material.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import festim as F
+import numpy as np
+import ufl
+
+test_mesh = F.Mesh1D(vertices=np.array([0.0, 1.0, 2.0, 3.0, 4.0]))
+x = ufl.SpatialCoordinate(test_mesh.mesh)
```
Comment (RemDelaporteMathurin) [inline test/test_material.py:6]: This isn't used in the file

File: test/test_material.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+def test_define_diffusion_coefficient():
+    """Test that the diffusion coefficient is correctly defined"""
+    T, D_0, E_D = 10, 1.2, 0.5
+
+    my_mat = F.Material(D_0=D_0, E_D=E_D)
+    D = my_mat.get_diffusion_coefficient(test_mesh.mesh, T)
+
+    D_analytical = D_0 * np.exp(-E_D / F.k_B / T)
+
+    assert float(D) == D_analytical
```
Comment (RemDelaporteMathurin) [inline test/test_material.py:18]: I would avoid using `==` for floating point comparison. 
Use `np.isclose` instead

File: test/test_permeation_problem.py [hunk 10]
```diff
@@ -134,4 +265,5 @@ def siverts_law(T, S_0, E_S, pressure):
 
 
 if __name__ == "__main__":
-    test_permeation_problem()
+    # test_permeation_problem()
+    test_permeation_problem_multi_volume()
```
Comment (RemDelaporteMathurin) [inline test/test_permeation_problem.py:269]: Can remove these lines altogether
```suggestion
    pass
```

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
Looks great! 

I have a few more comments

File: festim/subdomain/volume_subdomain.py [hunk 11]
```diff
@@ -37,9 +37,20 @@ def locate_subdomain_entities(self, mesh, vdim):
         Returns:
             entities (np.array): the entities of the subdomain
         """
+        self.check_borders_within_domain(mesh)
+
         entities = locate_entities(
-            mesh,
+            mesh.mesh,
```
Comment (RemDelaporteMathurin) [inline festim/subdomain/volume_subdomain.py:43]: We shouldn't to do the check here, so no need to change the mesh argument (note: if you wanted to change it, you shoud've changed the docstrings too)

File: festim/subdomain/volume_subdomain.py [hunk 12]
```diff
@@ -37,9 +37,20 @@ def locate_subdomain_entities(self, mesh, vdim):
         Returns:
             entities (np.array): the entities of the subdomain
         """
+        self.check_borders_within_domain(mesh)
```
Comment (RemDelaporteMathurin) [inline festim/subdomain/volume_subdomain.py:40]: This check doesn't belong here and should be done at the Problem level

**Comment by jhdark** [comment]
now contains a fix for issue #594

File: festim/hydrogen_transport_problem.py [hunk 13]
```diff
@@ -122,9 +134,21 @@ def define_markers_and_measures(self):
...
+        # check that subdomains are connected
+        if len(self.volume_subdomains) > 1:
+            all_borders = []
+            for vol in self.volume_subdomains:
+                for border in vol.borders:
+                    all_borders.append(border)
+            sorted_borders = np.sort(all_borders).reshape(int(len(all_borders) / 2), 2)
+            for i in range(0, len(sorted_borders) - 1):
+                if sorted_borders[i][1] != sorted_borders[i + 1][0]:
+                    raise ValueError("Subdomain borders don't match to each other ")
```
Comment (RemDelaporteMathurin) [inline festim/hydrogen_transport_problem.py]: I would have this in a separate method and check that the borders are within the mesh bounds at the same time!

File: test/test_subdomains.py [hunk 14]
```diff
@@ -40,9 +41,37 @@ def test_different_volume_ids():
     vol_subdomain_3 = F.VolumeSubdomain1D(
         id=vol_subdomains_ids[2], borders=[sub_dom_3[0], sub_dom_3[-1]], material=None
     )
-    my_model.subdomains = [vol_subdomain_1, vol_subdomain_2, vol_subdomain_3]
+    my_model.subdomains = [vol_subdomain_1, vol_subdomain_3, vol_subdomain_2]
```
Comment (RemDelaporteMathurin) [inline test/test_subdomains.py:44]: Why?

File: test/test_subdomains.py [hunk 15]
```diff
@@ -40,9 +41,37 @@ def test_different_volume_ids():
...
 
     my_model.define_markers_and_measures()
 
     for vol_id in vol_subdomains_ids:
         assert vol_id in np.array(my_model.volume_meshtags.values)
+
+
+def test_non_matching_volume_borders():
+    my_model = F.HydrogenTransportProblem()
```
Comment (RemDelaporteMathurin) [inline test/test_subdomains.py:53]:
```suggestion
def test_non_matching_volume_borders():
    """Checks that non-matching borders raise an error"""
    my_model = F.HydrogenTransportProblem()
```

File: test/test_subdomains.py [hunk 16]
```diff
@@ -40,9 +41,37 @@ def test_different_volume_ids():
...
+    sub_dom_2 = np.linspace(8e-05, 2e-04, num=5)
+    my_model.mesh = F.Mesh1D(np.unique(np.concatenate([sub_dom_1, sub_dom_2])))
+
+    vol_subdomains_ids = [1, 2]
+    vol_subdomain_1 = F.VolumeSubdomain1D(
+        id=vol_subdomains_ids[0], borders=[sub_dom_1[0], sub_dom_1[-1]], material=None
+    )
+    vol_subdomain_2 = F.VolumeSubdomain1D(
+        id=vol_subdomains_ids[1], borders=[sub_dom_2[0], sub_dom_2[-1]], material=None
+    )
```
Comment (RemDelaporteMathurin) [inline test/test_subdomains.py:65]: This can be simplified.

```suggestion
    my_model.mesh = F.Mesh1D(np.linspace(0, 10))

    vol_subdomain_1 = F.VolumeSubdomain1D(
        id=1, borders=[0, 2], material=None
    )
    vol_subdomain_2 = F.VolumeSubdomain1D(
        id=2, borders=[4, 10], material=None
    )
```

File: test/test_subdomains.py [hunk 17]
```diff
@@ -40,9 +41,37 @@ def test_different_volume_ids():
...
+    my_model.subdomains = [vol_subdomain_1, vol_subdomain_2]
+
+    with pytest.raises(ValueError):
+        my_model.define_markers_and_measures()
```
Comment (RemDelaporteMathurin) [inline test/test_subdomains.py:69]: I think we need a dedicated method for checking the subdomain borders.

File: test/test_subdomains.py [hunk 18]
```diff
@@ -40,9 +41,37 @@ def test_different_volume_ids():
...
+
+
+def test_borders_out_of_domain():
+    test_mesh = F.Mesh1D(vertices=np.linspace(0, 1, num=5))
+    sub_dom = F.VolumeSubdomain1D(id=1, borders=[1, 1.5], material=None)
+
+    with pytest.raises(ValueError):
+        sub_dom.locate_subdomain_entities(test_mesh, 1)
```
Comment (RemDelaporteMathurin) [inline test/test_subdomains.py:77]: See comment above.

Comment (RemDelaporteMathurin) [inline test/test_subdomains.py:77]: Let's try and test more edge cases here
```suggestion
@pytest.mark.parametrize("start,end", [(0.0, 0.5), (0.5, 1.0), (0.0, 1.5), (0.5, 1.5)])
def test_borders_out_of_domain(start, end):
    """tests that an error is raised when the borders don't match the boundaries of the mesh

    Args:
        start (float): the start of the subdomain
        end (float): the end of the subdomain
    """
    test_mesh = F.Mesh1D(vertices=np.linspace(0, 1, num=5))
    sub_dom = F.VolumeSubdomain1D(id=1, borders=[start, end], material=None)

    with pytest.raises(ValueError):
        sub_dom.locate_subdomain_entities(test_mesh, 1)
```

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
Some more comments.

If you're happy with the material implementation I would merge this without the boders check, and keep that for another PR as I have some comments on it.

Smaller PRs = better PRs

**Review by RemDelaporteMathurin — APPROVED** [review]
